### DISCUSSION TITLE
In process metrics

### DISCUSSION THREAD

**Discussion by pescuma** [metrics-rs/metrics#353] [General]
Is it possible to store the metrics in memory and use them inside the process itself? I'd like to use this to show some metrics while running some processor heavy tasks in a cli (and at the same time send those to Prometheus if the user wants).

**Comment by tobz** [comment]
## Recording metrics in-memory

It's possible to use something like [`DebuggingRecorder`](https://docs.rs/metrics-util/latest/metrics_util/debugging/struct.DebuggingRecorder.html) to achieve that. You would create the recorder, use `DebuggingRecorder::snapshotter` to get a `Snapshotter`, which is a handle to take a snapshot of metrics, and then install the recorder.

One thing to note is that `DebuggingRecorder` takes a lock when describing and registering new metrics, so if you're using the macros in a dynamic way -- such as `increment_counter!("my_metric", &some_var_with_dynamic_labels)` -- you'd be "registering" the counter every time that call was made, which would go through the locking path.

This lock is also accessed to take a snapshot of the recorder, so the insert performance, as well as the potential contention when taking snapshots, _may_ be noticeable at high metrics emission rates if you're emitting many "dynamic" metrics.

## Also sending metrics to Prometheus

You could utilize [layers](https://docs.rs/metrics-util/latest/metrics_util/layers/index.html) to achieve this, specifically the [`Fanout`](https://docs.rs/metrics-util/latest/metrics_util/layers/struct.FanoutBuilder.html) layer, which lets you fanout metrics to one or more recorders.

You would create the `DebuggingRecorder` and `PrometheusRecorder` recorders and add them both to `FanoutBuilder`. You'd then build the `Fanout` layer, which could be via `metrics::set_boxed_recorder(Box::new(your_fanout))`.

**Reply by pescuma** [reply]
Thanks, I'll try.
